File: scripts/add-lang-km-spans.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Adds .lang-km spans after .lang-en spans missing a paired .lang-km sibling."""

from __future__ import annotations

import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent / "src"

STATIC_KM = {
    "Done": "រួចរាល់",
    "Foundation": "មូលដ្ឋាន",
    "Beginner": "អ្នកចាប់ផ្តើម",
    "Intermediate": "កម្រិតមធ្យម",
    "General knowledge": "ចំណេះទូទៅ",
    "Why it matters": "ហេតុអ្វីសំខាន់",
    "Prerequisites": "តម្រូវការជាមុន",
    "Related Cloudflare products": "ផលិតផល Cloudflare ពាក់ព័ន្ធ",
    "On this hub": "នៅលើ hub",
    "Suggested tutorials": "Tutorial ណែនាំ",
    "Suggested exercise": "លំហាត់ណែនាំ",
    "Common newbie mistakes": "កំហុសថ្មីៗញឹកញាប់",
    "Sources": "ប្រភព",
    "Mark as complete": "សម្គាល់ថារួច",
    "Complete": "រួចរាល់",
    "Topics": "ប្រធានបទ",
    "Hub links": "តំណ hub",
    "Suggested deployment tutorials": "Tutorial deploy ណែនាំ",
    "Recommended products": "ផលិតផលណែនាំ",
    "Exercises": "លំហាត់",
    "Week outcome": "លទ្ធផលសប្តាហ៍",
    "Mark week complete": "សម្គាល់សប្តាហ៍រួច",
    "From zero": "ចាប់ផ្តើមពីសូន្យ",
    "Basic": "មូលដ្ឋាន",
    "Technical": "បច្ចេកទេស",
    "Mixed": "ចម្រុះ",
    "Duration": "រយៈពេល",
    "Starting level": "កម្រិតចាប់ផ្តើម",
    "Primary track": "Track ចម្បង",
    "Outcome": "លទ្ធផល",
    "View roadmap": "មើល roadmap",
    "Learning progress": "វឌ្ឍនភាពសិក្សា",
    "Reset": "Reset",
    "Progress overview": "ទិដ្ឋភាពវឌ្ឍនភាព",
    "Export JSON": "Export JSON",
    "Content Roadmap": "Content Roadmap",
    "Active role roadmap": "Roadmap តួនាទីសកម្ម",
    "Progress by role": "វឌ្ឍនភាពតាមតួនាទី",
    "Continue learning": "បន្តសិក្សា",
    "Roadmap progress": "វឌ្ឍនភាព roadmap",
    "Start this roadmap": "ចាប់ផ្តើម roadmap",
    "Continue": "បន្ត",
    "Glossary": "Glossary",
    "Checklist": "Checklist",
    "Quiz": "Quiz",
    "Keep learning": "បន្តសិក្សា",
    "Final outcome": "លទ្ធផលចុងក្រោយ",
    "Built by": "Built by",
    "Hosted on Cloudflare Pages": "Hosted on Cloudflare Pages",
}

LABEL_KM = {
    "Foundation": "មូលដ្ឋាន",
    "Beginner": "អ្នកចាប់ផ្តើម",
    "Intermediate": "កម្រិតមធ្យម",
    "From zero": "ចាប់ផ្តើមពីសូន្យ",
    "Basic": "មូលដ្ឋាន",
    "Technical": "បច្ចេកទេស",
    "Mixed": "ចម្រុះ",
}

SPAN_RE = re.compile(
    r'(<span (?:class|className)="[^"]*\blang-en\b[^"]*">)([\s\S]*?)(</span>)'
    r'(?!\s*\n?\s*<span (?:class|className)="[^"]*\blang-km\b)',
    re.MULTILINE,
)

LEVEL_LABEL_RE = re.compile(
    r"(\w+:\s*\{\s*vi:\s*'[^']*',\s*en:\s*'([^']*)'\s*\})"
)

INLINE_OBJ_RE = re.compile(
    r"\{\s*vi:\s*('(?:\\'|[^'])*'|\"(?:\\\"|[^\"])*\")"
    r"([^}]*?)en:\s*('(?:\\'|[^'])*'|\"(?:\\\"|[^\"])*\")(\s*)\}",
    re.MULTILINE,
)


def to_km_expression(expr: str) -> str:
    e = expr.strip()
    if "Km" in e:
        return e

    e = re.sub(
        r"([\w.?[\]'\"]+)\.(\w+)En(\s*\?\?\s*\1\.\2Vi)",
        r"\1.\2Km ?? \1.\2En\3",
        e,
    )
    e = re.sub(
        r"([\w.]+)\.(\w+)En\?\.(\[[^\]]+\])\s*\?\?\s*",
        r"\1.\2Km?.\3 ?? \1.\2En?.\3 ?? ",
        e,
    )

    m = re.fullmatch(r"(\w+)\.en", e)
    if m:
        return f"{m.group(1)}.km ?? {m.group(1)}.en"

    m = re.fullmatch(r"(\w+)\.titleEn", e)
    if m:
        return f"{m.group(1)}.titleKm ?? {m.group(1)}.titleEn"

    m = re.search(r"getContentTopicTitle\(([^,]+),\s*'en'\)", e)
    if m:
        arg = m.group(1)
        return f"getContentTopicTitle({arg}, 'km') ?? getContentTopicTitle({arg}, 'en')"

    m = re.search(r"getTutorialDisplayTitle\(([^,]+),\s*'en'\)", e)
    if m:
        arg = m.group(1)
        return f"getTutorialDisplayTitle({arg}, 'km') ?? getTutorialDisplayTitle({arg}, 'en')"

    return e


def static_km(text: str) -> str:
    trimmed = text.strip()
    return STATIC_KM.get(trimmed, trimmed)


def add_km_spans(src: str) -> str:
    def repl(m: re.Match[str]) -> str:
        open_tag, inner, close_tag = m.group(1), m.group(2), m.group(3)
        offset = m.start()
        line_start = src.rfind("\n", 0, offset) + 1
        indent_match = re.match(r"(\s*)", src[line_start:offset])
        indent = indent_match.group(1) if indent_match else "      "

        trimmed = inner.strip()
        if trimmed.startswith("{") and trimmed.endswith("}"):
            km_inner = "{" + to_km_expression(trimmed[1:-1]) + "}"
        else:
            km_inner = static_km(trimmed)

        open_km = re.sub(r"\blang-en\b", "lang-km", open_tag)
        return f"{open_tag}{inner}{close_tag}\n{indent}{open_km}{km_inner}{close_tag}"

    return SPAN_RE.sub(repl, src)


def add_km_to_level_labels(src: str) -> str:
    def repl(m: re.Match[str]) -> str:
        full, en_val = m.group(1), m.group(2)
        if "km:" in full:
            return full
        km = LABEL_KM.get(en_val, en_val)
        return full[:-1] + f", km: '{km}' }}"

    return LEVEL_LABEL_RE.sub(repl, src)


def add_km_to_inline_objects(src: str) -> str:
    def repl(m: re.Match[str]) -> str:
        match = m.group(0)
        if re.search(r"\bkm\s*:", match):
            return match
        vi_part, middle, en_part, tail = m.group(1), m.group(2), m.group(3), m.group(4)
        quote = en_part[0]
        en_val = en_part[1:-1]
        km_val = STATIC_KM.get(en_val, en_val)
        escaped = km_val.replace("\\", "\\\\").replace(quote, f"\\{quote}")
        return f"{{ vi: {vi_part}{middle}en: {en_part}, km: {quote}{escaped}{quote}{tail}}}"

    return INLINE_OBJ_RE.sub(repl, src)


def main() -> None:
    updated = 0
    skip_files = {"LangText.astro"}
    for path in sorted(ROOT.rglob("*")):
        if path.suffix not in {".astro", ".tsx"}:
            continue
        if path.name in skip_files:
            continue
        src = path.read_text(encoding="utf-8")
        if "lang-en" not in src:
            continue
        orig = src
        src = add_km_to_level_labels(src)
        src = add_km_to_inline_objects(src)
        src = add_km_spans(src)
        if src != orig:
            path.write_text(src, encoding="utf-8")
            updated += 1
            print(f"updated {path.relative_to(ROOT)}")

    print(f"Done. {updated} files updated.")


if __name__ == "__main__":
    main()

File: scripts/add-lang-km-spans.mjs
/**
 * Adds .lang-km spans after .lang-en spans missing a paired .lang-km sibling.
 * Also adds km fields to inline { vi, en } objects when missing.
 */
import { readFileSync, writeFileSync, readdirSync, statSync } from 'node:fs';
import { join, extname } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = join(fileURLToPath(new URL('.', import.meta.url)), '..', 'src');

const STATIC_KM = {
  Done: 'រួចរាល់',
  Foundation: 'មូលដ្ឋាន',
  Beginner: 'អ្នកចាប់ផ្តើម',
  Intermediate: 'កម្រិតមធ្យម',
  'General knowledge': 'ចំណេះទូទៅ',
  'Why it matters': 'ហេតុអ្វីសំខាន់',
  Prerequisites: 'តម្រូវការជាមុន',
  'Related Cloudflare products': 'ផលិតផល Cloudflare ពាក់ព័ន្ធ',
  'On this hub': 'នៅលើ hub',
  'Suggested tutorials': 'Tutorial ណែនាំ',
  'Suggested exercise': 'លំហាត់ណែនាំ',
  'Common newbie mistakes': 'កំហុសថ្មីៗញឹកញាប់',
  Sources: 'ប្រភព',
  'Mark as complete': 'សម្គាល់ថារួច',
  Complete: 'រួចរាល់',
  Topics: 'ប្រធានបទ',
  'Hub links': 'តំណ hub',
  'Suggested deployment tutorials': 'Tutorial deploy ណែនាំ',
  'Recommended products': 'ផលិតផលណែនាំ',
  Exercises: 'លំហាត់',
  'Week outcome': 'លទ្ធផលសប្តាហ៍',
  'Mark week complete': 'សម្គាល់សប្តាហ៍រួច',
  'From zero': 'ចាប់ផ្តើមពីសូន្យ',
  Basic: 'មូលដ្ឋាន',
  Technical: 'បច្ចេកទេស',
  Mixed: 'ចម្រុះ',
  Duration: 'រយៈពេល',
  'Starting level': 'កម្រិតចាប់ផ្តើម',
  'Primary track': 'Track ចម្បង',
  Outcome: 'លទ្ធផល',
  'View roadmap': 'មើល roadmap',
  'Learning progress': 'វឌ្ឍនភាពសិក្សា',
  Reset: 'Reset',
  'Progress overview': 'ទិដ្ឋភាពវឌ្ឍនភាព',
  'Export JSON': 'Export JSON',
  'Content Roadmap': 'Content Roadmap',
  'Active role roadmap': 'Roadmap តួនាទីសកម្ម',
  'Progress by role': 'វឌ្ឍនភាពតាមតួនាទី',
  'Continue learning': 'បន្តសិក្សា',
  'Roadmap progress': 'វឌ្ឍនភាព roadmap',
  'Start this roadmap': 'ចាប់ផ្តើម roadmap',
  Continue: 'បន្ត',
  Glossary: 'Glossary',
  Checklist: 'Checklist',
  Quiz: 'Quiz',
  Stage: 'ដំណាក់កាល',
  'After this stage you will': 'បន្ទាប់ពីដំណាក់កាលនេះ',
  'Role-based roadmaps': 'Roadmap តាមតួនាទី',
  'Keep learning': 'បន្តសិក្សា',
  'Final outcome': 'លទ្ធផលចុងក្រោយ',
  'Built by': 'Built by',
  'Hosted on Cloudflare Pages': 'Hosted on Cloudflare Pages',
};

function walk(dir, files = []) {
  for (const name of readdirSync(dir)) {
    const p = join(dir, name);
    if (statSync(p).isDirectory()) {
      if (!name.includes('node_modules')) walk(p, files);
    } else if (['.astro', '.tsx'].includes(extname(p))) {
      files.push(p);
    }
  }
  return files;
}

function toKmExpression(expr) {
  let e = expr.trim();
  if (/\bKm\b/.test(e)) return e;

  // fooEn ?? fooVi (same prefix)
  e = e.replace(
    /([\w.?[\]'"]+)\.(\w+)En(\s*\?\?\s*\1\.\2Vi)/g,
    '$1.$2Km$3 ?? $1.$2En$3',
  );

  // fooEn?.[i] ?? bar
  e = e.replace(
    /([\w.]+)\.(\w+)En\?\.(\[[^\]]+\])\s*\?\?\s*/g,
    '$1.$2Km?.$3 ?? $1.$2En?.$3 ?? ',
  );

  // label.en (LocalizedString)
  if (/^(\w+)\.en$/.test(e)) {
    const m = e.match(/^(\w+)\.en$/);
    return `${m[1]}.km ?? ${m[1]}.en`;
  }

  // t.titleEn
  if (/^(\w+)\.titleEn$/.test(e)) {
    const m = e.match(/^(\w+)\.titleEn$/);
    return `${m[1]}.titleKm ?? ${m[1]}.titleEn`;
  }

  // getContentTopicTitle(id, 'en')
  if (/getContentTopicTitle\(([^,]+),\s*'en'\)/.test(e)) {
    const m = e.match(/getContentTopicTitle\(([^,]+),\s*'en'\)/);
    return `getContentTopicTitle(${m[1]}, 'km') ?? getContentTopicTitle(${m[1]}, 'en')`;
  }

  // getTutorialDisplayTitle(preview, 'en')
  if (/getTutorialDisplayTitle\(([^,]+),\s*'en'\)/.test(e)) {
    const m = e.match(/getTutorialDisplayTitle\(([^,]+),\s*'en'\)/);
    return `getTutorialDisplayTitle(${m[1]}, 'km') ?? getTutorialDisplayTitle(${m[1]}, 'en')`;
  }

  return e;
}

function staticKm(text) {
  const trimmed = text.trim();
  if (STATIC_KM[trimmed]) return STATIC_KM[trimmed];
  return trimmed;
}

function addKmSpans(src) {
  const spanRe =
    /(<span (?:class|className)="[^"]*\blang-en\b[^"]*">)([\s\S]*?)(<\/span>)(?!\s*\n?\s*<span (?:class|className)="[^"]*\blang-km\b)/g;

  return src.replace(spanRe, (full, open, inner, close, offset) => {
    const lineStart = src.lastIndexOf('\n', offset) + 1;
    const indent = src.slice(lineStart, offset).match(/^(\s*)/)?.[1] ?? '      ';

    const trimmed = inner.trim();
    let kmInner;

    if (trimmed.startsWith('{') && trimmed.endsWith('}')) {
      const expr = trimmed.slice(1, -1);
      kmInner = `{${toKmExpression(expr)}}`;
    } else {
      kmInner = staticKm(trimmed);
    }

    const openAttrs = open.replace(/\blang-en\b/g, 'lang-km');
    return `${open}${inner}${close}\n${indent}${openAttrs}${kmInner}${close}`;
  });
}

function addKmToInlineObjects(src) {
  // Single-line { vi: '...', en: '...' } without km
  return src.replace(
    /\{\s*vi:\s*('(?:\\'|[^'])*'|"(?:\\"|[^"])*")([^}]*?)en:\s*('(?:\\'|[^'])*'|"(?:\\"|[^"])*")(\s*)\}/g,
    (match, viPart, middle, enPart, tail) => {
      if (/\bkm\s*:/.test(match)) return match;
      const enVal = enPart.slice(1, -1);
      const kmVal = STATIC_KM[enVal] ?? enVal;
      const kmQuote = enPart[0];
      const escaped = kmVal.replace(/\\/g, '\\\\').replace(new RegExp(kmQuote, 'g'), `\\${kmQuote}`);
      return `{ vi: ${viPart}${middle}en: ${enPart}, km: ${kmQuote}${escaped}${kmQuote}${tail}}`;
    },
  );
}

function addKmToLevelLabels(src) {
  const labelMap = {
    Foundation: 'មូលដ្ឋាន',
    Beginner: 'អ្នកចាប់ផ្តើម',
    Intermediate: 'កម្រិតមធ្យម',
    'From zero': 'ចាប់ផ្តើមពីសូន្យ',
    Basic: 'មូលដ្ឋាន',
    Technical: 'បច្ចេកទេស',
    Mixed: 'ចម្រុះ',
  };

  return src.replace(
    /(\w+:\s*\{\s*vi:\s*'[^']*',\s*en:\s*'([^']*)'\s*\})/g,
    (match, full, enVal) => {
      if (/km:/.test(match)) return match;
      const km = labelMap[enVal] ?? enVal;
      return full.replace(/\}\s*$/, `, km: '${km}' }`);
    },
  );
}

let updated = 0;
for (const file of walk(root)) {
  let src = readFileSync(file, 'utf8');
  if (!src.includes('lang-en')) continue;

  const orig = src;
  src = addKmToLevelLabels(src);
  src = addKmToInlineObjects(src);
  src = addKmSpans(src);

  if (src !== orig) {
    writeFileSync(file, src, 'utf8');
    updated++;
    console.log('updated', file.replace(root, ''));
  }
}
console.log(`Done. ${updated} files updated.`);
